linq4j/src/main/java/org/apache/calcite/linq4j/TopNHeap.java

```diff
@@ -0,0 +1,242 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.linq4j;
+
+import org.apache.calcite.linq4j.function.Function1;
+
+import java.util.Comparator;
+
+/**
+ * Implementation of a stable binary heap with a fetch and an offset.
+ * Stable means that if two items are considered equal,
+ * they will appear in the same order as they were offered to the heap.
+ *
+ * @param <TSource> type of the element that will be added to the heap
+ * @param <TKey> type of the key, which the comparator will use for comparisons
+ */
+public class TopNHeap<TSource, TKey> {
+  static final int MAX_INIT_ARRAY_SIZE = 1024;
+  static final int ROOT = 1;
+
+  private final Function1<TSource, TKey> keyFn;
+  private final Comparator<TKey> cmp;
+  private final int fetch;
+  private final int offset;
+  private final int maxSize;
+
+  TKey headKey = null;
+  int size = 0;
+  int time = -Integer.MIN_VALUE;
+
+  /** Heap with 1-based index, heap[0] is not used */
+  TSource[] heap;
+  /** Stores the order of arrival of the elements in the heap */
+  int[] order;
```

Comment:
I guess this field exists to ensure stability?

If so, you should state that stability is one of the goals of this class. Also, add tests for stability.

Is there any other way to achieve stability? An extra array (and element swaps at each stage) seems pretty expensive.

Reply:
Yes, it is for stability. Another possibility is to wrap the elements in another Object, which contains the order. The array is probably cheaper. There seems to be a more sophisticated data structure described [here](https://cstheory.stackexchange.com/questions/593/is-there-a-stable-heap), but it is much more difficult to implement, and I doubt that it is faster in practice.

Reply:
OK, so stability is a requirement of this class. That's fine. (We don't strictly need it for SQL, but it makes things more predictable.) Make it explicit in the description of the class.

The implementation is not very expensive, considering.